Allow loading mods from version-specified directory i.e. mods/rift/1.13/

k-jiang opened the pull request:
Let Rift to try loading mods form the `mods/"version"-Rift` folder. If empty, fall back to "mods/" instead.
p.s. I removed the "final" flag from variable `modsDir`, not sure if there are any consequences.

k-jiang committed:
Allow loading mods from version-specified directory i.e. mods/1.13-Rift

0f-0b commented on src/main/java/org/dimdev/riftloader/RiftLoader.java:
```diff
@@ -49,6 +50,17 @@ public void load(boolean isClient) {
         side = isClient ? Side.CLIENT : Side.SERVER;
 
+        // test if mods/"version"-Rift contains any .jar files
+        modsVersionSpecifiedDir.mkdir();
```
Why create a new <code>mods/_version_-Rift</code> while you can just check if the directory exists?

k-jiang replied:
Well, there are two reasons:
1. If folder `mods/_version_-Rift` does not exists, the checking will throw an exception. I just want to keep it simple
2. A special folder for Rift is better to separate between Rift's and forge's mods (or any other APIs). Since lots of ppl will just install multiple types of API together without a second thought. And guys like me would like to play different versions at the same time. So this new folder will make life much more easier without dealing too much with conflicts.

I'm still looking for solution to automate the `_version_` part. Either pull from `RealmsSharedConstants.VERSION_STRING` like you addressed in #43 . Or coded with `build.gradle` "replace" function. What's your suggestion?

sjx233 commented:
See #43.

ZombieHDGaming commented on src/main/java/org/dimdev/riftloader/RiftLoader.java:
```diff
@@ -29,7 +29,8 @@
 
-    public final File modsDir = new File(Launch.minecraftHome, "mods");
+    public File modsDir = new File(Launch.minecraftHome, "mods");
+    public final File modsVersionSpecifiedDir = new File(Launch.minecraftHome, "mods/1.13-Rift");
```
Since we are talking about making Rift mods at least cross 1.13 compatible, I'd suggest just making the folder Rift and making an exception for if the folder exists, search there. And please don't have the folder just created, just check to see if it's existing already.

k-jiang replied:
OK let me think a little bit....
I agree that cross compatibility is a very good idea. So maybe create a folder called `RiftMods` under `.minecraft/`?

Here is the rough draft. For backward compatibility, Rift's mod will load by this order:
1.  Create the directory called `RiftMods`, so ppl know where to put their mods
2. Try to see if the folder `RiftMods/_version_` exists, if there are any mods inside, load them
3. If not, try to see if there are any mods in `RiftMods/`, load from there
4. If not, load mods from `mods/` just like the old days

What do you think?

ZombieHDGaming replied:
What? No, the way it works is fine, just modify what you have from `1.13-Rift` to just `Rift` and just rather than straight up creating the folder, just add an if statement in the search to see if it exists, and if it does, search it.

k-jiang committed:
Don't create the "mods/1.13-Rift" folder by default.

k-jiang committed:
Automate minecraft version with mods directory

k-jiang committed:
Load mods from "mods/rift/" and its version-specified subdirectory
create "mods/rift/" folder by default
Rift's mods should be inside "mods/rift/" or "mods/rift/_version_", no more "mods/_version_-Rift/"
load mods from classpath, "mods/rift/_version_", "mods/rift/" and "mods/"
classpath mods are now loaded and logged separately to avoid confusion

k-jiang commented:
OK, I modify the loading process, now the logic is:
1. Load classpath mods, count and log it
2. Check if there are directory named "mods/rift/1.13/". If so, load jar mods from it, count, and log it
3. Create and load jar mods from "mods/rift/", count and log it
4. Create(even though step 3 already created) and load jar mods from "mods/", count and log it

The reason why I chose lower-case for "rift" is to eliminate any case-sensitive scenario and confusion, just like forge.
And users can create their own "1.13" sub-directory if they want. But "rift" will always be created so people know where to organize their mods, making it fool-proof (at least a little bit).

Any ideas?

0f-0b commented on src/main/java/org/dimdev/riftloader/RiftLoader.java:
```diff
@@ -29,7 +29,9 @@
     private static final Logger log = LogManager.getLogger("RiftLoader");
 
-    public final File modsDir = new File(Launch.minecraftHome, "mods");
+    public File modsDir = new File(Launch.minecraftHome, "mods");
```
```suggestion
    public final File modsDir = new File(Launch.minecraftHome, "mods");
```
The `final` flag.

0f-0b commented on src/main/java/org/dimdev/riftloader/RiftLoader.java:
```diff
@@ -49,19 +51,32 @@ public void load(boolean isClient) {
+            }
+        }
+        // load mods from folder mods/rift/, create the folder if not exists
+        findJarMods(modsRiftDir,true);
```
```suggestion
        findJarMods(modsRiftDir, true);
```

0f-0b commented on src/main/java/org/dimdev/riftloader/RiftLoader.java:
```diff
@@ -49,19 +51,32 @@ public void load(boolean isClient) {
+        // load mods from folder mods/rift/, create the folder if not exists
+        findJarMods(modsRiftDir,true);
+        // load mods from mods/, create the folder if not exists
+        findJarMods(modsDir,true);
```
```suggestion
        findJarMods(modsDir, true);
```

0f-0b commented on src/main/java/org/dimdev/riftloader/RiftLoader.java:
```diff
@@ -94,9 +109,22 @@ private void findMods(File modsDir) {
+    private void findJarMods(File modsDir) {
+        this.findJarMods(modsDir, false);
+    }
+    private void findJarMods(File modsDir, Boolean createDir) {
```
```suggestion

    private void findJarMods(File modsDir, boolean createDir) {
```

0f-0b commented on src/main/java/org/dimdev/riftloader/RiftLoader.java:
```diff
@@ -94,9 +109,22 @@ private void findMods(File modsDir) {
-        modsDir.mkdirs();
+        if (createDir) modsDir.mkdirs();
         for (File file : modsDir.listFiles()) {
             if (!file.getName().endsWith(".jar")) continue;
```
```suggestion
             if (!file.getName().toLowerCase().endsWith(".jar")) continue;
```

0f-0b commented on src/main/java/org/dimdev/riftloader/RiftLoader.java:
```diff
@@ -94,9 +109,22 @@ private void findMods(File modsDir) {
+        int modCount = modInfoMap.size();
         log.info("Searching for mods in " + modsDir);
-        modsDir.mkdirs();
+        if (createDir) modsDir.mkdirs();
```
```suggestion
        if (createDir) modsDir.mkdirs();
        else if (!(modsDir.exists() && modsDir.isDirectory())) return;
```
And this should be before `int modCount = modInfoMap.size();`

0f-0b commented on src/main/java/org/dimdev/riftloader/RiftLoader.java:
```diff
@@ -49,19 +51,32 @@ public void load(boolean isClient) {
+        // load mods from classpath
+        findClasspathMods();
+        // test if mods/rift/"version"/ contains any .jar files, then load from it
+        if (modsVersionSpecifiedDir.exists() && modsVersionSpecifiedDir.isDirectory()) {
```
Just `findJarMods(modsVersionSpecifiedDir);` and remove the tests, since they will be done in `findJarMods`.

0f-0b commented on src/main/java/org/dimdev/riftloader/RiftLoader.java:
```diff
@@ -29,7 +29,9 @@
-    public final File modsDir = new File(Launch.minecraftHome, "mods");
+    public File modsDir = new File(Launch.minecraftHome, "mods");
+    public final File modsVersionSpecifiedDir = new File(Launch.minecraftHome, "mods/rift/@MCVERSION@");
+    public final File modsRiftDir = new File(Launch.minecraftHome, "mods/rift");
```
```suggestion
    public final File modsRiftDir = new File(modsDir, "rift");
```

0f-0b commented on src/main/java/org/dimdev/riftloader/RiftLoader.java:
```diff
@@ -29,7 +29,9 @@
 
-    public final File modsDir = new File(Launch.minecraftHome, "mods");
+    public File modsDir = new File(Launch.minecraftHome, "mods");
+    public final File modsVersionSpecifiedDir = new File(Launch.minecraftHome, "mods/rift/@MCVERSION@");
```
```suggestion
    public final File modsVersionSpecifiedDir = new File(modsRiftDir, "@MCVERSION@");
```
And put this after `modsRiftDir` to resolve errors.

0f-0b reviewed (changes requested):
Suggest changes for readability and performance.

0f-0b committed:
Update src/main/java/org/dimdev/riftloader/RiftLoader.java

Fix case sensitive.

Co-Authored-By: k-jiang <k@kjiang.net>

0f-0b committed:
Apply suggestions from code review

Improve readability.

Co-Authored-By: k-jiang <k@kjiang.net>

k-jiang committed:
Improve loading logic, by @sjx233
